src/test/code-sample.test.ts

```diff
@@ -0,0 +1,169 @@
+/**
+ * @license
+ * Copyright (c) 2020 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {assert} from '@esm-bundle/chai';
+import '../lib/code-sample.js';
+import {CodeSampleElement} from '../lib/code-sample.js';
+import {html, render} from 'lit-html';
+import {UpdatingElement} from 'lit-element';
+import {CodeMirrorEditorElement} from '../lib/codemirror-editor.js';
+import {CodeSampleProjectElement} from '../lib/code-sample-project.js';
+
+describe('code-sample', () => {
+  let container: HTMLDivElement;
+
+  beforeEach(() => {
+    container = document.createElement('div');
+    document.body.appendChild(container);
+  });
+
+  afterEach(() => {
+    container.remove();
+  });
+
+  it('is registered', () => {
+    assert.instanceOf(document.createElement('code-sample'), CodeSampleElement);
+  });
+
+  const pierce = async (...selectors: string[]) => {
+    let node = document.body;
+    for (const selector of selectors) {
+      const result = (node.shadowRoot ?? node).querySelector(selector);
+      assert.instanceOf(result, HTMLElement);
+      if ((result as UpdatingElement).updateComplete) {
+        await (result as UpdatingElement).updateComplete;
+      }
+      node = result as HTMLElement;
+    }
+    return node;
+  };
+
+  const assertPreviewContains = async (text: string) => {
+    const iframe = (await pierce(
+      'code-sample',
+      'code-sample-preview',
+      'iframe'
+    )) as HTMLIFrameElement;
+    await new Promise((resolve) => {
+      const listener = () => {
+        iframe.removeEventListener('load', listener);
+        resolve();
+      };
+      iframe.addEventListener('load', listener);
+    });
+    // TODO(aomarks) Chromium and Webkit both fire iframe "load" after the
+    // contentDocument has actually loaded, but Firefox fires it before. Why is
+    // that? If not for that, we wouldn't need to poll here.
+    await new Promise((resolve) => {
+      const check = () => {
+        if (iframe.contentDocument?.body?.textContent?.includes(text)) {
+          resolve();
+        } else {
+          setTimeout(check, 10);
```

Comment:
should you limit the number of times this will retry?

Reply:
Mocha has its own test timeout, so I don't think we need another one. But it does seem like a good idea to kill the polling if the mocha timeout has been reached. Added a variable to stop calling `setTimeout` if `teardown` has been called.